src/translate.py

```diff
@@ -431,29 +446,78 @@ def get_stack_info(
+            callee_saved_words.append(stack_offset)
         elif (
             inst.mnemonic in ["sw", "swc1", "sdc1"]
             and destination in SAVED_REGS
             and isinstance(inst.args[1], AsmAddressMode)
             and inst.args[1].rhs.register_name == "sp"
+            and destination not in info.callee_save_reg_locations
         ):
             # Initial saving of callee-save register onto the stack.
-            info.callee_save_reg_locations[destination] = inst.args[1].lhs_as_literal()
-
-    # Find the region that contains local variables. It is above saved registers
-    # and the return address, if those exist. If they don't, the local variables
-    # can lie directly at the bottom of the stack.
-    info.local_vars_region_bottom = 0
+            stack_offset = inst.args[1].lhs_as_literal()
+            info.callee_save_reg_locations[destination] = stack_offset
+            if inst.mnemonic == "sdc1":
+                callee_saved_doubles.append(stack_offset)
+            else:
+                callee_saved_words.append(stack_offset)
+        elif (
+            inst.mnemonic in ["lw", "lwc1", "ldc1"]
+            and isinstance(inst.args[1], AsmAddressMode)
+            and inst.args[1].rhs.register_name == "sp"
+        ):
+            info.subroutine_arg_top = update_min(
+                info.subroutine_arg_top, inst.args[1].lhs_as_literal()
+            )
+        elif (
+            inst.mnemonic == "addiu"
+            and inst.args[1] == Register("sp")
+            and isinstance(inst.args[2], AsmLiteral)
+            and inst.args[2].value < info.allocated_stack_size
+        ):
+            info.subroutine_arg_top = update_min(
+                info.subroutine_arg_top, inst.args[2].value
+            )
 
     if not info.is_leaf:
-        info.local_vars_region_bottom = info.return_addr_location + 4
+        # Compute the bounds of the callee-saved register region, including padding
+        callee_saved_words.sort()
+        callee_saved_doubles.sort()
+        bottom = callee_saved_words[0]
+
+        # Both GCC & IDO save registers in two blocks: one for word-sized registers,
+        # followed by one for doubles. The word-sized block is also padded to a
+        # multiple of 8 bytes. Check that there are no gaps in either of these blocks.
+        top = bottom
+        for offset in callee_saved_words:
+            assert offset == top, DecompFailure(
+                f"Gap in callee-saved word stack region. "
+                f"Saved words: {callee_saved_words}, "
+                f"saved doubles: {callee_saved_doubles}, "
+                f"gap at: {offset} != {top}."
+            )
+            top += 4
+        # Padding: round up to multiple of 8
+        if (top % 8) == 4:
+            top += 4
+        for offset in callee_saved_doubles:
+            if offset < callee_saved_words[-1]:
+                # This offset is fake: it's a local variable; exclude it
```

Comment:
can this happen? I'd expect it not to because all saved registers must be saved before being used in any other way

Reply:
Good catch - I think when I added the `and destination not in info.callee_save_reg_locations` check, it removed the issue I was originally seeing.

...but when I went to verify that, I discovered that IDO actually puts the doubles *below* the words. So, I'll update this logic to work in either orientation.